src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
I'm not a big fan of method bloat but I also don't like the idea of having a bunch of extra memory fences. It's probably not that big of a deal since most of the users will be on x86, but it's worth thinking about.

I don't think that the verbosity is much of a problem in writing compare_exchange, and I'd be fine with removing compare_and_swap completely in favor of compare_exchange.